### ISSUE TITLE
Alternator: missing ItemCount field in return of DescribeTable

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#7550]
The return from `DescribeTable` is missing the `ItemCount` field. According to DynamoDB documentation, `ItemCount` should have:

> The number of items in the specified table. DynamoDB updates this value approximately every six hours. Recent changes might not be reflected in this value. 

Unfortunately, there is no obvious way to implement this feature efficiently in Scylla. We have a probabilistic mechanism for estimating the number of partitions in a set of sstables, but `ItemCount` is about individual rows, not partitions. Moreover it's not clear to me (we'll need to find an answer to this) whether a probabilistic, approximate count, is good enough - or if this must be an **accurate** count of items (albeit only true to six hours ago).

**Comment by ptrsmrn** [comment]
I think that just performing `SELECT COUNT(*) from alternator_cf.cf (ALLOW FILTERING)` every 6h and storing the count in cf's "metadata" (for tablets and indexes) is a brave idea but also may be enough.
Amazon leaves plenty of room for error when sizing the table:
> DynamoDB updates this value approximately every six hours. Recent changes might not be reflected in this value

So the count is updated "approximately" each 6h, perhaps 4 times a day. That plus "Recent changes might not be reflected in this value", especially the use of unprecise "recent" word, which blurrs the accuracy of this measurement, makes me believe we could get away with just the simple mechanism I described at the beginning. 
But even if I'm wrong, the above is probably better than nothing.

**Comment by nuivall** [comment]
If it's inaccurate anyway we can try first using probabilistic mechanism and query nodes on describe request. I assume we can somehow combine per node result to get total count. It's a bit wasteful to do periodic jobs, potentially scanning the whole table just for the purpose of a rarely used API call.

**Comment by ScyllaPiotr** [comment]
Could we implement this only for tablets?
Could we associate an `ItemCount` value with each tablet?
Could we update the `ItemCount` value only on each compaction of a tablet, accepting the fact that we will not be as up-to-date as DynamoDB with the numbers?
On each `DescribeTable` query could we fetch the value from each of the table's tablet, deduplicate tablet replicas and then sum the numbers?

**Comment by nyh** [comment]
> Could we implement this only for tablets?

Yes, at this point of the development process - where LWT on tablets is already being worked so Alternator will then fully support tablets - I think that if you have an idea that only works for tablets, you can assume tables.

> Could we associate an `ItemCount` value with each tablet?

Yes, we could.

> Could we update the `ItemCount` value only on each compaction of a tablet, accepting the fact that we will not be as up-to-date as DynamoDB with the numbers?

Please remember that compaction is not always, or even at all, "major compaction" involving all the sstables of the tablet. So it's not obvious how to "combine" the count you get from the output of the partial compaction, with the count you had before for the tablet.

I had a different idea of something we can do with compaction instead of trying to maintain a count of items: we could try to maintain an estimate of the "density" of items per gigabyte - e.g., imagine that during some random compaction we read 1GB of sstables, and output 10,000 rows. We then have an estimate that 1GB of on-disk files contain about 10,000 rows. Perhaps if we do different compactions of different sizes and average them, we'll get some average number N of rows per 1GB, and then at any time we can just look at the amount of disk used and multiply it by N to get a good estimate of the current number of rows. But I'm worried that this idea will work well for LCS but less for STCS because its space amplification can get fairly large and lead to tens of percents of discrepancies on the estimated number of rows. Maybe with AI we can learn how an sstable's age and compaction history leads to space amplification and can correct this estimate :-) In short, it doesn't sound simple.

**Comment by ScyllaPiotr** [comment]
> I had a different idea of something we can do with compaction instead of trying to maintain a count of items: we could try to maintain an estimate of the "density" of items per gigabyte - e.g., imagine that during some random compaction we read 1GB of sstables, and output 10,000 rows. We then have an estimate that 1GB of on-disk files contain about 10,000 rows. Perhaps if we do different compactions of different sizes and average them, we'll get some average number N of rows per 1GB, and then at any time we can just look at the amount of disk used and multiply it by N to get a good estimate of the current number of rows. But I'm worried that this idea will work well for LCS but less for STCS because its space amplification can get fairly large and lead to tens of percents of discrepancies on the estimated number of rows. Maybe with AI we can learn how an sstable's age and compaction history leads to space amplification and can correct this estimate :-) In short, it doesn't sound simple.

Indeed, this idea would work well with little amount of overwriting or frequent compacting, basically only when the data density calculated **after** compaction resembles the average data density (regardless how long ago the SSTables were compacted). Otherwise, like you wrote, more (overly?) advanced mechanisms would be needed.

So maybe another idea:
- count/estimate writes that are not overwrites and subtract deletions of existing items
  - check the bloom filter if it is a write or an update, and if a deletion is called for an existing item, factor in the false positive rate of the bloom filter
- monitor only some percentage of operations (like 1%) and scale the result (like 100x)

With storing of the relevant metadata per SSTable, your idea above (counting the actual number of items on each compaction, plus storing the amount of data churned by compaction) could easily enhance the calculations I propose.

**Comment by radoslawcybulski** [comment]
If i remember correctly each SSTable has a bloom filter for keys to allow fast skip. We can use this bloom filter to estimate how much keys in two SSTables overlap (roughly). Knowing number of rows in single SSTable we could estimate total rows in two SSTable's together. I'm not sure how to do that for more than 2 SSTables, i seem to vaguely remember from my math days, that we'd have to calculate bloom filter overlap for each possible pair and do some magic math with it. In general - given number of rows per SSTable we should be able to calculate number of rows in set of SSTables.

We could deal with tombstones by keeping per shard in table ratio of mutation to deletion calculated after each compaction in tag value. Or randomly decode some random keys from few randomly selected SSTable's and calculate ratio ourselves.

**Comment by nyh** [comment]
> If i remember correctly each SSTable has a bloom filter for keys to allow fast skip. We can use this bloom filter to estimate how much keys in two SSTables overlap (roughly).

Right, and we actually do something similar - it's called "cardinality estimator". See http://www.datastax.com/dev/blog/improving-compaction-in-cassandra-with-cardinality-estimation. It estimates how many **partitions** the result of the compaction will have. This works, to make a long story short, by intersecting bloom filters (sort of) - see http://blog.notdot.net/2012/09/Dam-Cool-Algorithms-Cardinality-Estimation for the very interesting tricks that make this work.

But - the Bloom filter, and the cardinaility estimator, is about the number of **partitions**, not about the number of rows ("items" in DynamoDB speak) that we neeed in this issue...

**Comment by radoslawcybulski** [comment]
True, i forgot that. I guess we'd use average rows per partition? With sufficiently large tables those should be close to truth.

**Comment by ScyllaPiotr** [comment]
I did some more research with the help of [some contemporary tool](https://deepwiki.com/search/why-does-alternator-do-readbef_586dfdf0-e5df-4f8d-9826-a65bf2631e2b#4) (it handled my typos really well too). Here's another idea - make each tablet belonging to a specific table return an estimate of the number of elements it holds, calculated as a sum of:
- number of rows in each SSTable belonging to the tablet, `rows_count` is part of SSTable's statistics metadata;
- number of **partitions** in each memtable, accessible using the getter `partition_count()`

The idea has the following issues effecting in imprecise results:
- the SSTables of a tablet almost certainly contain duplicated data, so adding up all rows in all SSTables will likely result in a number **higher** than the actual number of persisted rows (that's what the compaction is for)
- counting partitions in memtables instead of rows will likely result in a number **lower** than the actual number of rows

Still, there are several advantages to the approach:
- the required calculations are very quick, it's O(No_of_memtables + No_of_SSTables)
- no [additional costly operations](https://github.com/scylladb/scylladb/issues/7550#issuecomment-2457331098) are required
- I'd claim the relative precision of the estimation should increase with the age of the DB (with more data persisted)
- the numbers will change on par with the inflow of the data (as long as more partitions are created within memtables), which will make good impression on the users who could watch/show the data frequently

**Comment by mykaul** [comment]
@ScyllaPiotr - what about tombstones? Do they somehow get accounted for? How?

**Comment by ScyllaPiotr** [comment]
The `rows_count` field of SSTable represents the total count of all rows without a (straightforward) way to subtract the dead/tombstoned rows. I'd say this is an extreme case of the _SSTables of a tablet almost certainly contain duplicated data_ bullet above, since row creation inside one earlier SSTable and deletion of the same row inside one later SSTable would be counted as 2 rows, while in reality there is 0 rows.

**Comment by radoslawcybulski** [comment]
Ratio of tombstones to duplicated rows is dependand on usage pattern.
We can estimate row overlap using bloom filter trick.
We can calculate tombstone to value ratio as we go (ie as inserts / deletions happen) and write it as a tag every X minutes.

Assuming TOTAL_ROW_COUNT, ROW_OVERLAP and TOMBSTONE_TO_VALUE values final value would be:

MAX(0, (TOTAL_ROW_COUNT / ROW_OVERLAP) / TOMBSTONE_TO_VALUE - 1)

EDIT: looks decent, nice job @ScyllaPiotr !

**Comment by ScyllaPiotr** [comment]
Although I did not get that to the letter, I believe I understand the general idea (which is to increase precision by watching actual SSTables). What I would suggest is to do this probing:
- without running any additional processes, so for instance while doing compaction
- later, in v2 of the solution, since I suspect these are not straightforward to do
